store_io.go

```diff
@@ -0,0 +1,600 @@
+	cidBuf := pool.Get(len(cidKey))
+	copy(cidBuf, cidKey)
+
+	rows, err := dbbs.dbPool.Query(
+		context.TODO(),
+		fmt.Sprintf(
+			`
+			SELECT %s
+				FROM fil_common_base.datablocks
+			WHERE
+				cid = $1::BYTEA
+					AND
+				size IS NOT NULL
+			`,
+			StoredBlocksInflatorSelection,
+		),
+		cidBuf,
+	)
+
+	pool.Put(cidBuf)
+
+	if err != nil {
+		if rows != nil {
+			rows.Close()
+		}
+		return nil, err
+	}
+
+	sbs, _, err := dbbs.InflateDbRows(rows, (aType == GET), false)
+
+	switch {
+	case err != nil:
+		return nil, err
+	case len(sbs) == 0:
+		return nil, nil
+	case len(sbs) > 1:
+		log.Panicf("impossibly(?) retrieved %d rows for a single unique cid '%s'", len(sbs), rootCid.String())
+		return nil, nil
+	default:
+		dbbs.noteAccess(*sbs[0].dbBlockOrdinal, time.Now(), aType)
+		return sbs[0], nil
+	}
+}
+
+// InflateDbRows transforms result of
+// 	`SELECT %StoredBlocksInflatorSelection% FROM fil_common_base.datablocks ...`
+// to a set of StoredBlock's, which in turn are the structures implementing
+// github.com/ipfs/go-block-format.Block() Before returning Close()s the
+// supplied pgx.Rows()
+func (dbbs *PgBlockstore) InflateDbRows(rows pgx.Rows, eagerUnpack bool, skipCaching bool) (blocks []*StoredBlock, bytesCached int64, err error) {
+	defer rows.Close()
+
+	ret := make([]*StoredBlock, 0)
+
+	for rows.Next() {
+		var sb StoredBlock
+		var size *int32
+		var cidBytes, dbContentBytes []byte
+
+		err := rows.Scan(&cidBytes, &sb.dbBlockOrdinal, &size, &sb.dbContentEncoding, &dbContentBytes)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		// the SQL query pulled a placeholder: skip
+		if size == nil {
+			continue
+		}
+		sb.size = *size
+
+		if sb.cid, err = cid.Cast(cidBytes); err != nil {
+			return nil, 0, err
+		}
+
+		switch {
+
+		case sb.cid.Prefix().MhType == multihash.IDENTITY:
+			dmh, _ := multihash.Decode(sb.cid.Hash()) // no error checks - we just made this CID
+			sb.blockContent = dmh.Digest
+
+		case sb.dbContentEncoding == nil:
+			return nil, 0, fmt.Errorf(
+				"unexpectedly unspecified (nil) content encoding type for block_ordinal %d (%s)",
+				sb.dbBlockOrdinal,
+				sb.cid.String(),
+			)
+
+		case *sb.dbContentEncoding == 0:
+			// *MUST* make a copy, otherwise the LRU will trap the Pgx backing array
+			// https://github.com/jackc/pgx/issues/845#issuecomment-705550012
+			sb.blockContent = append(
+				make([]byte, 0, len(dbContentBytes)),
+				dbContentBytes...,
+			)
+
+		case *sb.dbContentEncoding == 1:
+
+			// put the magic-prefix back: the copy serves double-duty avoiding
+			// https://github.com/jackc/pgx/issues/845#issuecomment-705550012
+			sb.dbDeflatedContent = append(append(
+				make([]byte, 0, len(zstdMagic)+len(dbContentBytes)),
+				zstdMagic...,
+			), dbContentBytes...)
+
+			// we know we will need the content soon: trigger decompression in the background
+			// do so without a limiter: if we starve the CPU for a bit - so be it
+			if eagerUnpack {
+				go sb.inflatedContent(false)
```

Comment:
This seems wasteful considering the attention to small optimizations elsewhere. What if the fetch of all rows fails just before the last one? There's no way to stop these goroutines. Also the number of goroutines is unbounded with unknown memory allocation for each.

Reply:
You are correct, remnant from an ancient iteration.
Removed from function call and placed at the right spot here: https://github.com/filecoin-project/go-bs-postgres-chainnotated/commit/8b1f2e9296c4346ca7647889cde5d2ecae123f8e#diff-97588be80ee1efc9827cef3de764f82d5ede8bdc7ead07d610268e49f84e70d6L129-R182